Bulk Attachment Download Frontend

frankchn opened the pull request:
## Problem
This pull request adds the modal and other frontend features necessary to enable bulk attachment downloading.

## Solution
This adds a dropdown and modal and otherwise uses the existing csvDownload and attachment download backends.

Closes #16

liangyuanruo reviewed (commented):
code looks great, let me test this on staging later

frankchn commented:
Added P-Queue based implementation to restrict concurrent downloads to 1:

![image](https://user-images.githubusercontent.com/691628/100533878-250ba700-31be-11eb-85fb-5a8a4469d39e.png)

liangyuanruo commented on src/public/modules/forms/helpers/decryption.worker.js:
```diff
@@ -150,10 +152,10 @@ async function decryptIntoCsv(data) {
-        downloadBlob = await downloadAndDecryptAttachmentsAsZip(
-          attachmentDownloadUrls,
-          secretKey,
+        let downloadPromise = queue.add(() =>
+          downloadAndDecryptAttachmentsAsZip(attachmentDownloadUrls, secretKey),
         )
+        downloadBlob = await downloadPromise
```
```suggestion
        downloadBlob = await queue.add(() =>
          downloadAndDecryptAttachmentsAsZip(attachmentDownloadUrls, secretKey),
        )
```

liangyuanruo commented on src/public/modules/forms/services/submissions.client.factory.js:
```diff
@@ -270,7 +270,12 @@ function SubmissionsFactory(
 
           // Create a pool of decryption workers
-          const numWorkers = $window.navigator.hardwareConcurrency || 4
+          let numWorkers = $window.navigator.hardwareConcurrency || 4
+          if (downloadAttachments) {
+            // If we are downloading attachments, we restrict the number of threads
+            // to one to limit resource usage on the client's browser.
+            numWorkers = 1
+          }
```
```suggestion
          // If we are downloading attachments, we restrict the number of threads
          // to one to limit resource usage on the client's browser.
          const numWorkers = downloadAttachments ? 1 : ($window.navigator.hardwareConcurrency || 4)
```

liangyuanruo reviewed (approved):
lgtm - do we have any more UI changes to make?

liangyuanruo commented on src/public/modules/forms/helpers/decryption.worker.js:
```diff
@@ -150,9 +152,8 @@ async function decryptIntoCsv(data) {
-        downloadBlob = await downloadAndDecryptAttachmentsAsZip(
-          attachmentDownloadUrls,
-          secretKey,
+        downloadBlob = await queue.add(() =>
+          downloadAndDecryptAttachmentsAsZip(attachmentDownloadUrls, secretKey),
```
to add try-catch block to isolate attachment failure from CSV decryption failure

frankchn replied:
Added

liangyuanruo reviewed (changes requested):
**Action items**
1. To display separate wording in modal if there are no attachments to download.
2. To improve messaging on the modal to emphasiSe that it is beta, experimental, and requires the user to have ample disk space and be mentally prepared for a large number of files to be triggered.
3. To add Google Analytics monitoring (#762)

syan-syan commented:
Chrome
- pre-download modal should be page center not flushed to the top
- Export dropdown options should not disconnect when user scrolls down
![Uploading Screenshot 2020-12-01 at 3.26.48 PM.png…]()

frankchn commented:
- Added separate wording for no attachments (wording to be improved later).
- Added beta tags to dropdown and modal.
- Moved modal to center of screen
- Resolved issue of dropdown disconnecting
- Added google analytics monitoring for failed downloads (Closes #762)

![image](https://user-images.githubusercontent.com/691628/101273535-16d40280-374b-11eb-8230-d172dc3e684e.png)

liangyuanruo commented:
code looks good, @syan-syan if you could confirm that the wording is OK?

liangyuanruo commented on src/public/modules/forms/admin/views/download-all-attachments.client.modal.html:
```diff
@@ -0,0 +1,32 @@
+      Download attachments for {{ responsesCount }} responses
+      <span class="beta-icon">beta</span>
+    </div>
+    <div ng-if="responsesCount > 0" class="field-description">
+      <p>
+        Downloading many attachments may take a long time. The attachments from
+        each response will be downloaded as a separate ZIP file. Start now, or
+        adjust your date range before proceeding.
+      </p>
+    </div>
+    <div ng-if="responsesCount === 0" class="field-description">
+      <p>
+        The date range you selected does not contain any responses. Please
+        select a date range containing responses and try again.
+      </p>
+    </div>
```
@syan-syan the wording is here

liangyuanruo commented:
@frankchn I just had a quick discussion with @syan-syan, and we'd like to clear one last hurdle before releasing next week which is that the submission ID and timestamp aren't being included in the downloaded CSV if any part of the web worker decryption fails. It's important because, without it, the user can't provide us any info or contact us for help. 

could you help us to add that in? the other form fields can simply be left blank.

frankchn commented:
Here are the current designs. They should be pretty close to the Zeplin references.

![image](https://user-images.githubusercontent.com/691628/102028226-f7963000-3d5d-11eb-9a96-2687cfdd2855.png)

![image](https://user-images.githubusercontent.com/691628/102028242-10064a80-3d5e-11eb-8295-544d0755f787.png)

![image](https://user-images.githubusercontent.com/691628/102028264-24e2de00-3d5e-11eb-8e8f-2bf21a7a1525.png)

![image](https://user-images.githubusercontent.com/691628/102028277-388e4480-3d5e-11eb-8ea9-e6a1621ac84e.png)

Also, added the submission ID and worker timestamp into the CSV as an additional column regardless of whether the submission succeeds or fails.

```
Expected total responses,2
Total decrypted responses,2
Error count,0
Unverified response count,0
Reference number,Timestamp,test attachment 1,test text 123,Download Status
5fd6a89fc9b363002aca6c6d,14 Dec 2020 07:49:51 AM,abcdef.pdf,STATUS123456,Success
5fd6a75fc9b363002aca6c69,14 Dec 2020 07:44:31 AM,ghijkl.jpg,,Success
```

liangyuanruo commented:
thanks @frankchn ... i believe @syan-syan is envisioning a slightly different format for the CSV. let's firm it up at sync-up tomorrow.

frankchn commented:
Changed CSV downloads to be the following format:

```
Expected total responses,2
Success count,2
Error count,0
Unverified response count,0
See download status column for download errors
Reference number,Timestamp,Attachment Download Status,test attachment 1,test text 123
5fd6a89fc9b363002aca6c6d,14 Dec 2020 07:49:51 AM,Success,abcdef.pdf,STATUS123456
5fd6a75fc9b363002aca6c69,14 Dec 2020 07:44:31 AM,Success,def.jpg,123456
```

liangyuanruo commented:
> ```
> Expected total responses,2
> Success count,2
> Error count,0
> Unverified response count,0
> See download status column for download errors
> Reference number,Timestamp,Attachment Download Status,test attachment 1,test text 123
> 5fd6a89fc9b363002aca6c6d,14 Dec 2020 07:49:51 AM,Success,abcdef.pdf,STATUS123456
> 5fd6a75fc9b363002aca6c69,14 Dec 2020 07:44:31 AM,Success,def.jpg,123456
> ```

@frankchn not sure if @syan-syan told you, but could we change `Attachment Download Status` column to the more generic `Download Status` that would also show the status for decryption failure etc.?

liangyuanruo commented on src/public/modules/forms/services/submissions.client.factory.js:
```diff
@@ -297,7 +302,12 @@ function SubmissionsFactory(
+              if (csvRecord.status === 'ATTACHMENT_ERROR') {
+                attachmentErrorCount++
+                errorCount++
+                // accumulate dataset anyway, since the download is otherwise OK
```
we also want to accumulate the submission ID and timestamp, if the decryption fails

frankchn replied:
Changed

liangyuanruo commented on src/public/modules/forms/helpers/decryption.worker.js:
```diff
@@ -150,10 +152,27 @@ async function decryptIntoCsv(data) {
+          csvRecord.status = 'ATTACHMENT_ERROR'
+          downloadStatus.answer = 'Failed'
+        }
+        csvRecord.submissionData.record.unshift(downloadStatus)
```
is it possible that `record` is uninitialised at this point, causing the worker to crash?

frankchn replied:
True, resolved with 93d52fe

frankchn commented:
Design changes:

https://www.figma.com/file/ta5tjjBqphML8LPRkmxD53/DesignMaster_v2?node-id=928%3A45

liangyuanruo reviewed (changes requested):
Code is good; merge pending copy changes.

frankchn commented:
![image](https://user-images.githubusercontent.com/691628/103396964-dfa70600-4aea-11eb-838c-d110252b57cd.png)
![image](https://user-images.githubusercontent.com/691628/103396973-eafa3180-4aea-11eb-8b31-5b5d70761aef.png)
![image](https://user-images.githubusercontent.com/691628/103397029-3d3b5280-4aeb-11eb-914e-75ec7a71f588.png)

syan-syan commented:
@liangyuanruo just made a tiny commit changing copy

liangyuanruo commented on src/public/modules/forms/admin/views/view-responses.client.view.html:
```diff
@@ -105,11 +105,44 @@
-            is-loading="vm.csvDownloading"
-          ></export-button-component>
+          <div class="btn-group" uib-dropdown>
+            <button
+              class="btn-custom btn-small"
+              aria-haspopup="true"
+              aria-expanded="false"
+              uib-dropdown-toggle
+              id="btn-export-dropdown"
+            >
+              <span class="hidden-xs" ng-hide="vm.csvDownloading">Export</span>
+              <span class="hidden-xs" ng-show="vm.csvDownloading"
+                >Downloading</span
+              >
+              <span class="caret"></span>
+            </button>
+            <ul
+              class="dropdown-menu dropdown-menu-right"
+              id="export-dropdown-options"
+              uib-dropdown-menu
+              role="menu"
+              aria-labelledby="btn-export-dropdown"
+            >
+              <li role="menuitem">
+                <a
+                  id="btn-export-dropdown-responses"
+                  ng-click="vm.exportCsv(false)"
+                  style="cursor: pointer"
+                  >CSV only</a
+                >
+              </li>
+              <li role="menuitem">
+                <a
+                  ng-click="vm.confirmSubmissionCountsBeforeDownload()"
+                  style="cursor: pointer"
+                  >CSV and attachments <span class="beta-icon">beta</span></a
+                >
+              </li>
+            </ul>
+          </div>
```
we should probably refactor this HTML and associated CSS into a new `<dropdown-menu-button>` component for reuse.

liangyuanruo reviewed (approved):
lgtm, let's test on UAT later

syan-syan commented:
copy changes

- pre-download modal infobox: "Downloading many attachments can be an intensive operation" remove the full stop after this line
- pre-download modal infobox: "ensure network connectivity is stable"
- all modals: header should say "responses and attachments"
